### PR TITLE
Shen WenQiang's DevOps Homework

### PR DESCRIPTION
**Pull request by sven0219**
I completed all the tasks in the test, but I am not familiar with all the skills in the test.
For example the skills required in Task 4 and Task 6 are relatively new to me.
Also, some playbooks and kubernetes config files reference files from other existing projects.

Sven Shen
swq1227@gmail.com

### CODE REVIEW

**Comment by kaleocheng** [comment]
@Sven219 Thanks for your time to finish those tasks.  I will start to review them today

File: Task2/roles/deploy_server/tasks/configure.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+- name: API | install pip
+  yum:
```
Comment (kaleocheng) [inline Task2/roles/deploy_server/tasks/configure.yml:2]: better to use `package`, so you can run this on non-Redhat server as well. and don't need this then https://github.com/Wiredcraft/test-devops/pull/31/files#diff-19b901b4ed86df0713f50231d8178c6ab952bfda2b6bfbf8e0f9aa12262d70c4R4

File: Task2/roles/initialization/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# core set
+core_dump: "yes"
+
+# directory set
+soft: /data/soft
+app: /data/app
+data: /data/data
+logs: /data/logs
+publish: /data/publish
+unarchive: /data/unarchive
```
Comment (kaleocheng) [inline Task2/roles/initialization/defaults/main.yml]: could you help to explain what's  those folders for

File: Task2/roles/initialization/tasks/vimrc.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+---
+# file: vimrc.yml
+#
+- name: Vim | set vim encoding for it
+  copy:
+    src: vimrc
+    dest: /home/it/.vimrc
+    owner: it
```
Comment (kaleocheng) [inline Task2/roles/initialization/tasks/vimrc.yml]: I saw the `it` name in lots of files, better to make a var for it

File: Task3/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+Create an Elastic Compute Service (ECS) instance by using Terraform on Alibaba Cloud.
+# Environment
+- Terraform v1.1.9
+- alicloud provider plugin aliyun/alicloud 1.164.0
+
+
+# Prerequisites
+
+Before you begin, ensure that you have completed the following operations:
+- Prepare an Alibaba Cloud account and an AccessKey pair (AccessKey ID and AccessKey secret) to use Terraform. You can go to the [Security Management](https://usercenter.console.aliyun.com/?spm=a2c63.p38356.0.0.402519dbawMNu3#/manage/ak) page of the Alibaba Cloud console to create or view your AccessKey pair.
```
Comment (kaleocheng) [inline Task3/README.md]: just put the `https://usercenter.console.aliyun.com/` is enough

File: Task4/db.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/go-redis/redis"
+)
+
+var rdb *redis.Client
+
+// init redis config
+func initClient() (err error) {
+	config, err := getRedisConfig()
+	if err != nil {
+		log.Printf("Get config file error: %v", err)
```
Comment (kaleocheng) [inline Task4/db.go]: should raise the `err` out instead of logging only

File: Task4/db.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	_, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	_, err = rdb.Ping().Result()
+	return err
+}
+
+// get key from redis
+func redisGet(key string) string {
```
Comment (kaleocheng) [inline Task4/db.go]: suggest to return both `error` and `string`. e.g. the error from `initClient()`  has to be handled.

File: Task4/db.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	if err := initClient(); err != nil {
+		return ""
+	}
+	val, err := rdb.Get(key).Result()
+	if err == redis.Nil {
+		fmt.Println("name does not exist")
```
Comment (kaleocheng) [inline Task4/db.go]: usually we don't use `fmt.Println` directly in  prod code. it's more for debugging.  use logger instead

File: Task2/roles/initialization/defaults/main.yml (hunk 2 above)
Comment (sven0219) [reply Task2/roles/initialization/defaults/main.yml]:
```yaml
soft: /data/soft   #   Third-party software installation directory
app: /data/app   #    Software installation directory developed by yourself
data: /data/data  #  The data directory of the software, such as mysql data
logs: /data/logs  #    All logs storage directory
publish: /data/publish  #  Store some scripts used for publishing
unarchive: /data/unarchive  #  Store unarchived files
```
These are some of my own usage habits, but these directories are not used in this task, so I will delete them.

File: Task2/roles/deploy_server/tasks/configure.yml (hunk 1 above)
Comment (sven0219) [reply Task2/roles/deploy_server/tasks/configure.yml:2]: I didn't know there was a `package` module before this. now I know.

File: Task2/roles/initialization/tasks/vimrc.yml (hunk 3 above)
Comment (sven0219) [reply Task2/roles/initialization/tasks/vimrc.yml]: As you suggested, I set `it` to var.

**Comment by kaleocheng** [comment]
@Sven219 Thanks for your time on those tasks again. I don't have other question ATM.  will get back to you by email soon.   have a nice day.

**Pull request closed without merging**